### ISSUE TITLE
Does ivy-use-virtual-buffers just link to the bookmarks file?

### ISSUE THREAD

**Issue by DevJac** [abo-abo/swiper#2334]
I have enabled `ivy-use-virtual-buffers`, but all it seems to do is link to the bookmarks file. Meaning it adds only one suggestion to `counsel-switch-buffer`, and that one suggestion is `bookmarks`. If I select `bookmarks` in `counsel-switch-buffer`, it brings up the bookmark file as shown below:

```
;;;; Emacs Bookmark Format Version 1 ;;;; -*- coding: utf-8-emacs -*- 
;;; This format is meant to be slightly human-readable;
;;; nevertheless, you probably don't want to edit it.
;;; -*- End Of Bookmark File Format Version Stamp -*-
(("elisp-study"
 (front-context-string . "\n7 ‘car’, ‘cdr’,")
 (rear-context-string . "ening,  Up: Top\n")
 (position . 235006)
 (filename . "/usr/share/info/eintr")
 (info-node . "car cdr & cons")
 (handler . Info-bookmark-jump))
)
```

Is this behaving as intended? When I read that it includes bookmarks in the suggestions I was thinking it would actually list the names of bookmarks I've created in `counsel-switch-buffer` and that selecting a bookmark would take me to the selected bookmarks location.

**Comment by abo-abo** [comment]
Seems like a bug in your config. Can you reproduce this with `make plain`?

**Comment by DevJac** [comment]
```
;; The following line is a workaround for Emacs 26.2.
(setq gnutls-algorithm-priority "NORMAL:-VERS-TLS1.3")

(package-initialize)
(setq package-archives '(("gnu" . "https://elpa.gnu.org/packages/")
			 ("melpa" . "https://melpa.org/packages/")))

(setq package-selected-packages
      '(counsel))         ; suggested requirement of lsp

;;; theme config
(set-face-attribute 'default nil :height 160)

(ivy-mode 1)
(counsel-mode 1)
(setq ivy-use-virtual-buffers t)
```
This is my `init.el`. I started with a fresh `~/.emacs.d/ directory with only that `init.el` file and installed ivy/counsel and was able to reproduce the bug. I set a bookmark in my `init.el` file and then restarted Emacs and then `bookmarks` was an option in `ivy-switch-buffer`, when I selected it it shows my `bookmarks` file as originally described.

**Comment by abo-abo** [comment]
> What is make plain?

`make plain` is a shell command that produces a running Emacs with only `ivy` loaded:

```sh
git clone https://github.com/abo-abo/swiper/
cd swiper
make plain
```

When I ran `make plain` and <kbd>C-x b</kbd>, indeed `bookmarks` was shown in the list of the virtual buffers.

Which is correct, since the virtual buffers are bookmarks+recentf. All bookmarks were present. And the `bookmarks` file was added to the head of `recentf-list`, since it was the last visited file.

If you don't want recentf, you can set:

```elisp
(setq ivy-use-virtual-buffers 'bookmarks)
```

**Comment by DevJac** [comment]
Ok. I checked out that project and did `make plain`. I modified `targets/plain.el` and added `(setq ivy-use-virtual-buffers t)`. Then I closed Emacs and started it again with `make plain`.
I could see `bookmarks` in `ivy-switch-buffer`, and when I selected it it took me to the bookmarks file, which contained the following:
```
;;;; Emacs Bookmark Format Version 1 ;;;; -*- coding: utf-8-emacs -*- 
;;; This format is meant to be slightly human-readable;
;;; nevertheless, you probably don't want to edit it.
;;; -*- End Of Bookmark File Format Version Stamp -*-
(("a-test"
 (filename . "~/Code/elisp/swiper/ivy-test.el")
 (front-context-string . "(defvar ivy-expr")
 (rear-context-string . "'ffap-url-p)))\n\n")
 (position . 1641))
("org-config"
 (filename . "~/.emacs.d/init.el")
 (front-context-string . ";;; org config\n(")
 (rear-context-string . "Tree\"\n	\" WK\"))\n\n")
 (position . 1151))
)
```
As you can see, I have two bookmarks, named `org-config` and `a-test`. I couldn't get either bookmark to show up as an option in `ivy-switch-buffer`.

Is it expected that `org-config` and `a-test` (my bookmark names) should show up in `ivy-switch-buffer`?

**Comment by abo-abo** [comment]
> As you can see, I have two bookmarks, named `org-config` and `a-test`. I couldn't get either bookmark to show up as an option in `ivy-switch-buffer`.

They are supposed to show up as "ivy-test.el" and "init.el", /not/ by their bookmark names. Maybe that's the source of the confusion?

Because I just did a `make plain` test with a `~/.emacs.d/bookmarks` like this:

```elisp
;;;; Emacs Bookmark Format Version 1 ;;;; -*- coding: utf-8-emacs -*-
;;; This format is meant to be slightly human-readable;
;;; nevertheless, you probably don't want to edit it.
;;; -*- End Of Bookmark File Format Version Stamp -*-
(("test1"
  (filename . "~/git/gnu-emacs/README"))
 ("test2"
  (filename . "~/git/gnu-emacs/info/dir")))
```

And everything worked as expected.

**Comment by DevJac** [comment]
> They are supposed to show up as "ivy-test.el" and "init.el", /not/ by their bookmark names. Maybe that's the source of the confusion?
Yes, this was the point of confusion. I was expecting the actual bookmarks to show in switch buffer, not just the file that was bookmarked.

Thanks for clarifying that. You can close the issue now, unless you have something more you want to discuss.

If you're ever looking for new features to add, I would like if the actual bookmark **names** showed in `ivy-switch-buffer`. If you think it's a good idea, then I may look into adding it myself.

**Comment by abo-abo** [comment]
Thanks, closing the issue.

> If you're ever looking for new features to add, I would like if the actual bookmark names showed in ivy-switch-buffer. If you think it's a good idea, then I may look into adding it myself.

There's `counsel-bookmark` for that. When Ivy gets a feature for merging sources, it will be possible to easily merge `ivy-switch-buffer` and `counsel-bookmark`.

**Comment by cjauvin** [comment]
I've been using ivy for a while (thank you for it it's really great!), but yesterday I read the manual more carefully, and realized that my `ivy-use-virtual-buffers` (already set to `1`, which means it includes everything), should  make `switch-to-buffer` _also_ display bookmarks. As I was expecting to somehow see the bookmark names in the list, I fiddled for a while, trying to understand why it didn't work (I suspected `ivy-rich` for a while, among other theories). From this thread I seem to understand that what I have in mind does not exist.
